test/comps/test_qb.cc

```diff
@@ -15,115 +15,108 @@ class TestQB : public ::testing::Test
+        std::vector<T> S;
+        std::vector<T> U;
+        std::vector<T> VT;
+
+        QBTestData(int64_t m, int64_t n, int64_t k) :
+                    A(m * n, 0.0), 
+                    Q(m * k, 0.0), 
+                    B(k * n, 0.0), 
+                    B_cpy(k * n, 0.0), 
+                    A_hat(m * n, 0.0),
+                    A_k(m * n, 0.0),  
+                    A_cpy(m * n, 0.0),  
+                    A_cpy_2(m * n, 0.0),
+                    A_cpy_3(m * n, 0.0),
+                    s(n, 0.0),
+                    S(n * n, 0.0),
+                    U(m * n, 0.0),
+                    VT(n * n, 0.0)
+                    {}
+    };
+
+    template <typename T, typename RNG>
+    struct algorithm_objects {
+        RandLAPACK::PLUL<T> Stab;
+        RandLAPACK::RS<T, RNG> RS;
+        RandLAPACK::CholQRQ<T> Orth_RF;
+        RandLAPACK::RF<T> RF;
+        RandLAPACK::CholQRQ<T> Orth_QB;
+        RandLAPACK::QB<T> QB;
+
+        algorithm_objects(bool verbosity, 
+                            bool cond_check, 
+                            bool orth_check, 
+                            int64_t p, 
+                            int64_t passes_per_iteration, 
+                            RandBLAS::base::RNGState<RNG> state) :
+                            Stab(cond_check, verbosity),
+                            RS(Stab, state, p, passes_per_iteration, verbosity, cond_check),
+                            Orth_RF(cond_check, verbosity),
+                            RF(RS, Orth_RF, verbosity, cond_check),
+                            Orth_QB(cond_check, verbosity),
+                            QB(RF, Orth_QB, verbosity, orth_check)
+                            {}
+    };
+
+    template <typename T>
+    static void svd_and_copy_computational_helper(int64_t m, int64_t n, QBTestData<T>& all_data) {
+        
+        // Create a copy of the original matrix
+        blas::copy(m * n, all_data.A.data(), 1, all_data.A_cpy.data(), 1);
+        blas::copy(m * n, all_data.A.data(), 1, all_data.A_cpy_2.data(), 1);
+        blas::copy(m * n, all_data.A.data(), 1, all_data.A_cpy_3.data(), 1);
+
+        // Get low-rank SVD
+        lapack::gesdd(Job::SomeVec, m, n, all_data.A_cpy.data(), m, all_data.s.data(), all_data.U.data(), m, all_data.VT.data(), n);
+    }
+
     /// General test for QB:
     /// Computes QB factorzation, and checks:
     /// 1. A - QB
     /// 2. B - \transpose{Q}A
     /// 3. I - \transpose{Q}Q
     /// 4. A_k - QB = U_k\Sigma_k\transpose{V_k} - QB
     template <typename T, typename RNG>
-    static void test_QB2_low_exact_rank(int64_t m, int64_t n, int64_t k, int64_t p, int64_t block_sz, T tol, std::tuple<int, T, bool> mat_type, RandBLAS::base::RNGState<RNG> state) {
+    static void test_QB2_low_exact_rank(
+        int64_t m, 
+        int64_t n, 
+        int64_t k,  
+        int64_t block_sz, 
+        T tol,  
+        QBTestData<T>& all_data,
+        algorithm_objects<T, RNG>& all_algs) {
```

Comment:
I don't think you need ``algorithm_objects`` as an argument. It looks like it suffices to provide the ``QB`` object.

That said, I do like the idea of having a helper function that defines a QB object from the following parameters.
```
    bool verbosity, 
    bool cond_check, 
    bool orth_check, 
    int64_t p, 
    int64_t passes_per_iteration, 
    RandBLAS::base::RNGState<RNG> state
```

Reply:
A helper function does not work, same issue as we had yesterday.
Upon exist from such function, objects stop existing.
That is why a made a struct.

Reply:
Nested initialization should work.

```
QB(
    bool verbosity, 
    bool cond_check, 
    bool orth_check, 
    int64_t p, 
    int64_t passes_per_iteration, 
    RandBLAS::base::RNGState<RNG> state
) : RF_obj(
        RS_obj(
            Stab_obj(...),
            ....
        ),
        Orth_obj(...),...
    ),
    Orth_obj(...) {}
```
This assumes that the member associated with each class that needs an ``Alg`` object is called ``Alg_obj``, all the way from Orth to Stab.

This actually would amount to a custom constructor for the QB class.

Reply:
There is a design question related to this. For a given algorithm class, all subroutine algorithm objects have reference types. That is, if something were to change in the subroutine, this will be reflected in the main algorithm. I don't think we're using this feature anywhere at the moment, but it makes sense to be able to use it.
 Because of that, we cannot exactly do a nested decloration - neither GCC allows for it, nor it technically makes sense.

Reply:
So the only alternatives i see right now are the way i did things in the most recent commit, or just declaring all objects in TEST_F environment and passing the QB object into the test function.

Reply:
Ugh. This is frustrating. Okay, we'll go with your struct solution for now. I'll want to revisit this later.